fetch.js

```diff
@@ -73,60 +73,92 @@
     this._body = null
     this.bodyUsed = false
 
-    this.arrayBuffer = function() {
-      throw new Error('Not implemented yet')
-    }
+    if (has_responseType) {
+      this.arrayBuffer = function() {
+        var rejected = consumed(this)
+        if (rejected) {
+          return rejected
+        }
 
-    var blobSupport = (function() {
-      try {
-        new Blob();
-        return true
-      } catch(e) {
-        return false
+        var body = this._body
+        return new Promise(function(resolve, reject) {
+          var fileReader = new FileReader()
+          fileReader.readAsArrayBuffer(body)
+          fileReader.onloadend = function(e) {
+            if (e.target.error) {
+              reject(e.target.error)
+            } else {
+              resolve(e.target.result)
+            }
+          }
+        })
       }
-    })();
 
-    if (blobSupport) {
       this.blob = function() {
         var rejected = consumed(this)
-        return rejected ? rejected : Promise.resolve(new Blob([this._body]))
+        return rejected ? rejected : Promise.resolve(this._body)
+      }
+
+      this.text = function() {
+        var rejected = consumed(this)
+        if (rejected) {
+          return rejected
+        }
+
+        var body = this._body
+        return new Promise(function(resolve, reject) {
+          var fileReader = new FileReader()
+          fileReader.readAsText(body)
+          fileReader.onloadend = function(e) {
+            if (e.target.error) {
+              reject(e.target.error)
+            } else {
+              resolve(e.target.result)
+            }
+          }
+        })
+      }
+    } else {
+      this.text = function() {
+        var rejected = consumed(this)
+        return rejected ? rejected : Promise.resolve(this._body)
       }
     }
 
     if (self.FormData) {
       this.formData = function() {
-        var rejected = consumed(this)
-        return rejected ? rejected : Promise.resolve(decode(this._body))
+        return this.text().then(function(text) {
+          return new Promise(function(resolve, reject) {
+            try {
+              resolve(decode(text))
+            } catch (ex) {
+              reject(ex)
+            }
+          })
+        })
       }
     }
 
     this.json = function() {
-      var rejected = consumed(this)
-      if (rejected) {
-        return rejected
-      }
-
-      var body = this._body
-      return new Promise(function(resolve, reject) {
-        try {
-          resolve(JSON.parse(body))
-        } catch (ex) {
-          reject(ex)
-        }
+      return this.text().then(function(text) {
+        return new Promise(function(resolve, reject) {
+          try {
+            resolve(JSON.parse(text))
+          } catch (ex) {
+            reject(ex)
+          }
```

Comment:
You don't need to make a new Promise here.

Just do `this.text().then(function(text) { return JSON.parse(text) })`